File: 03. Comportamiento/01. Chain of Responsibility/01. chainOfResponsibility.js
/***

Imaginemos que estamos construyendo una aplicación de procesamiento de solicitudes de compra. 
Tenemos diferentes niveles de aprobación para las solicitudes de compra y queremos implementar 
un flujo de trabajo donde cada nivel de aprobación pueda tomar una decisión sobre la solicitud 
y pasarla al siguiente nivel si es necesario.

Aquí es donde el patrón Chain of Responsibility puede ser útil. Podemos utilizar el patrón 
Chain of Responsibility para construir una cadena de objetos que representan los diferentes 
niveles de aprobación y cada objeto en la cadena puede tomar una decisión sobre la solicitud 
y pasarla al siguiente objeto en la cadena si es necesario.

*/

// Comencemos definiendo nuestra clase base Approver que representa un nivel de aprobación:

class Approver {
    constructor() {
        this.nextApprover = null;
    }

    setNextApprover(approver) {
        this.nextApprover = approver;
    }

    processRequest(request) {
        throw new Error("El método processRequest() debe ser implementado por las subclases.");
    }
}

// A continuación, definimos nuestras clases concretas que extienden la clase base Approver:

class Manager extends Approver {
    processRequest(request) {
        if (request.amount <= 1000) {
            console.log("La solicitud de compra ha sido aprobada por el gerente.");
        } else if (this.nextApprover) {
            this.nextApprover.processRequest(request);
        } else {
            console.log("La solicitud de compra no puede ser aprobada.");
        }
    }
}

class Director extends Approver {
    processRequest(request) {
        if (request.amount <= 5000) {
            console.log("La solicitud de compra ha sido aprobada por el director.");
        } else if (this.nextApprover) {
            this.nextApprover.processRequest(request);
        } else {
            console.log("La solicitud de compra no puede ser aprobada.");
        }
    }
}

class CEO extends Approver {
    processRequest(request) {
        if (request.amount <= 10000) {
            console.log("La solicitud de compra ha sido aprobada por el CEO.");
        } else {
            console.log("La solicitud de compra no puede ser aprobada.");
        }
    }
}

// Ahora podemos utilizar el patrón Chain of Responsibility para procesar solicitudes de compra:

const manager = new Manager();
const director = new Director();
const ceo = new CEO();

manager.setNextApprover(director);
director.setNextApprover(ceo);

const request1 = { amount: 800 };
manager.processRequest(request1); // La solicitud de compra ha sido aprobada por el gerente.

const request2 = { amount: 3500 };
manager.processRequest(request2); // La solicitud de compra ha sido aprobada por el director.

const request3 = { amount: 15000 };
manager.processRequest(request3); // La solicitud de compra no puede ser aprobada.

/***

En este ejemplo, el patrón Chain of Responsibility nos permite construir una cadena de objetos 
Approver donde cada objeto puede tomar una decisión sobre la solicitud de compra y pasarla al 
siguiente objeto en la cadena si es necesario. Esto nos permite implementar un flujo de trabajo 
flexible y escalable para procesar solicitudes de compra.

*/

File: 03. Comportamiento/01. Chain of Responsibility/01. chainOfResponsibility.py
# Definimos nuestra clase base Approver que representa un nivel de aprobación
class Approver:
    def __init__(self):
        self.next_approver = None

    def set_next_approver(self, approver):
        self.next_approver = approver

    def process_request(self, request):
        raise NotImplementedError("El método process_request() debe ser implementado por las subclases.")

# Definimos nuestras clases concretas que extienden la clase base Approver
class Manager(Approver):
    def process_request(self, request):
        if request['amount'] <= 1000:
            print("La solicitud de compra ha sido aprobada por el gerente.")
        elif self.next_approver:
            self.next_approver.process_request(request)
        else:
            print("La solicitud de compra no puede ser aprobada.")

class Director(Approver):
    def process_request(self, request):
        if request['amount'] <= 5000:
            print("La solicitud de compra ha sido aprobada por el director.")
        elif self.next_approver:
            self.next_approver.process_request(request)
        else:
            print("La solicitud de compra no puede ser aprobada.")

class CEO(Approver):
    def process_request(self, request):
        if request['amount'] <= 10000:
            print("La solicitud de compra ha sido aprobada por el CEO.")
        else:
            print("La solicitud de compra no puede ser aprobada.")

# Utilizamos el patrón Chain of Responsibility para procesar solicitudes de compra
manager = Manager()
director = Director()
ceo = CEO()

manager.set_next_approver(director)
director.set_next_approver(ceo)

request1 = {'amount': 800}
manager.process_request(request1) # La solicitud de compra ha sido aprobada por el gerente.

request2 = {'amount': 3500}
manager.process_request(request2) # La solicitud de compra ha sido aprobada por el director.

request3 = {'amount': 15000}
manager.process_request(request3) # La solicitud de compra no puede ser aprobada.